src/sdk/pynni/nni/compression/torch/builtin_pruners.py

```diff
@@ -102,7 +182,168 @@ def compute_target_sparsity(self, config):
+        self.if_init_list = {}
+
+    def calc_mask(self, layer, config):
+        """
+        Calculate the mask of given layer
+
+        Parameters
+        ----------
+        layer : LayerInfo
+            the layer to instrument the compression operation
+        config : dict
+            layer's pruning config
+
+        Returns
+        -------
+        torch.Tensor
+            mask of the layer's weight
+
+        """
+
+        weight = layer.module.weight.data
+        op_name = layer.name
+        op_type = layer.type
+        assert op_type == 'Conv2d', 'FilterPruner only supports 2d convolution layer pruning'
+        if self.if_init_list.get(op_name, True):
+            kernels = weight.shape[0]
+            w_abs = weight.abs()
+            k = int(kernels * config['sparsity'])
+            if k == 0:
+                return torch.ones(weight.shape).type_as(weight)
+            w_abs_structured = w_abs.view(kernels, -1).sum(dim=1)
+            threshold = torch.topk(w_abs_structured.view(-1), k, largest=False).values.max()
+            mask = torch.gt(w_abs_structured, threshold)[:, None, None, None].expand_as(weight).type_as(weight)
+            self.mask_dict.update({op_name: mask})
+            self.if_init_list.update({op_name: False})
+        else:
+            mask = self.mask_dict[op_name]
+        return mask
+
+
+class SlimPruner(Pruner):
+    """
+    A structured pruning algorithm that prunes channels by pruning the weights of BN layers.
+    Zhuang Liu, Jianguo Li, Zhiqiang Shen, Gao Huang, Shoumeng Yan and Changshui Zhang
+    "Learning Efficient Convolutional Networks through Network Slimming", 2017 ICCV
+    https://arxiv.org/pdf/1708.06519.pdf
+
+    Attributes
+    ----------
+    mask_dict : dict
+        Dictionary for saving masks.
+
+    """
+
+    def __init__(self, model, config_list):
+        """
+        Initiate `FilterPruner` from `config_list`
+        config_list: supported keys:
+            - sparsity
+
+        Parameters
+        ----------
+        config_list : list
+            List of pruning configs
+
+        """
+
+        super().__init__(model, config_list)
+        self.if_init_list = {}
+        weight_list = []
+        config = self.config_list[0]
+        op_types = config.get('op_types')
+        op_names = config.get('op_names')
+        if op_types is not None:
+            assert op_types == ['BatchNorm2d'], 'SlimPruner only supports 2d batch normalization layer pruning'
+            for name, m in model.named_modules():
+                if type(m).__name__ == 'BatchNorm2d':
+                    weight_list.append(m.weight.data.clone())
+        else:
+            for name, m in model.named_modules():
+                if name in op_names:
+                    assert type(
+                        m).__name__ == 'BatchNorm2d', 'SlimPruner only supports 2d batch normalization layer pruning'
+                    weight_list.append(m.weight.data.clone())
```

Comment:
this part seems not correct, you should know the relation between `op_types` and `op_names`

Reply:
suggest to use `detect_modules_to_compress` to get all the to-be-compressed modules. this member func is in pr #1698

Reply:
I'll resolve this when pr 1698 merged

Reply:
I've resolved this, thanks.